### PR TITLE
JDK-8306584: Start of release updates for JDK 22

### PR DESCRIPTION
**Pull request by jddarcy**
Time to get JDK 22 underway...

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8306627](https://bugs.openjdk.org/browse/JDK-8306627) to be approved
- [x] Change requires CSR request [JDK-8306628](https://bugs.openjdk.org/browse/JDK-8306628) to be approved
- [x] Change requires CSR request [JDK-8306626](https://bugs.openjdk.org/browse/JDK-8306626) to be approved

### Issues
 * [JDK-8306584](https://bugs.openjdk.org/browse/JDK-8306584): Start of release updates for JDK 22 (**Bug** - P4)
 * [JDK-8306585](https://bugs.openjdk.org/browse/JDK-8306585): Add SourceVersion.RELEASE_22 (**Sub-task** - P4)
 * [JDK-8306586](https://bugs.openjdk.org/browse/JDK-8306586): Add source 22 and target 22 to javac (**Sub-task** - P4)
 * [JDK-8306627](https://bugs.openjdk.org/browse/JDK-8306627): Add SourceVersion.RELEASE_22 (**CSR**)
 * [JDK-8306626](https://bugs.openjdk.org/browse/JDK-8306626): Add source 22 and target 22 to javac (**CSR**)
 * [JDK-8306628](https://bugs.openjdk.org/browse/JDK-8306628): Start of release updates for JDK 22 (**CSR**)

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to [6047c2cb](https://git.openjdk.org/jdk/pull/13567/files/6047c2cb0529888393880eff82dea4ad9f41004e)
 * [Iris Clark](https://openjdk.org/census#iris) (@irisclark - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [376dbe26](https://git.openjdk.org/jdk/pull/13567/files/376dbe26957a6c438815f4db428e8aab67935076)
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**) ⚠️ Review applies to [376dbe26](https://git.openjdk.org/jdk/pull/13567/files/376dbe26957a6c438815f4db428e8aab67935076)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13567/head:pull/13567` \
`$ git checkout pull/13567`

Update a local copy of the PR: \
`$ git checkout pull/13567` \
`$ git pull https://git.openjdk.org/jdk.git pull/13567/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13567`

View PR using the GUI difftool: \
`$ git pr show -t 13567`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13567.diff">https://git.openjdk.org/jdk/pull/13567.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13567#issuecomment-1560090712)

### CODE REVIEW

**Comment by jddarcy** [comment]
Typical start-of-JDK-next changes. As usual, the bulk of the PR is the symbol file information to support "javac --release 21".

File: test/langtools/tools/javac/versions/Versions.java [hunk 1]
```diff
@@ -23,7 +23,7 @@
 
 /*
  * @test
- * @bug 4981566 5028634 5094412 6304984 7025786 7025789 8001112 8028545 8000961 8030610 8028546 8188870 8173382 8173382 8193290 8205619 8028563 8245147 8245586 8257453 8286035
+ * @bug 4981566 5028634 5094412 6304984 7025786 7025789 8001112 8028545 8000961 8030610 8028546 8188870 8173382 8173382 8193290 8205619 8028563 8245147 8245586 8257453 8286035 8306586
```
Comment (jonathan-gibbons) [inline test/langtools/tools/javac/versions/Versions.java:26]: At some point, this line should be wrapped

File: test/langtools/tools/javac/versions/Versions.java [hunk 2]
```diff
@@ -89,7 +89,8 @@ static enum SourceTarget {
         EIGHTEEN(false,  "62.0", "18", Versions::checksrc18),
         NINETEEN(false,  "63.0", "19", Versions::checksrc19),
         TWENTY(false,    "64.0", "20", Versions::checksrc20),
-        TWENTY_ONE(false,"65.0", "21", Versions::checksrc20);
+        TWENTY_ONE(false,"65.0", "21", Versions::checksrc20),
+        TWENTY_TWO(false,"66.0", "22", Versions::checksrc20);
```
Comment (jonathan-gibbons) [inline test/langtools/tools/javac/versions/Versions.java:93]: when do these get updated?

**Review by erikj79 — APPROVED** [review]

File: test/langtools/tools/javac/versions/Versions.java (hunk 2 above)
Comment (jddarcy) [reply test/langtools/tools/javac/versions/Versions.java:93]: Ah, when new non-preview language structures are added. With two new language features in JDK 21, looks like it is time for me to add another one; thanks.

**Review by irisclark — APPROVED** [review]

File: test/langtools/tools/javac/versions/Versions.java (hunk 2 above)
Comment (jddarcy) [reply test/langtools/tools/javac/versions/Versions.java:93]: Better to fix this omission in JDK 21; please see https://github.com/openjdk/jdk/pull/14229.

**Review by irisclark — APPROVED** [review]

File: test/langtools/tools/javac/versions/Versions.java (hunk 2 above)
Comment (jddarcy) [reply test/langtools/tools/javac/versions/Versions.java]: > Better to fix this omission in JDK 21; please see #14229.

Addressed by merge of JDK-8309134 into JDK 22; thanks.

**Review by dholmes-ora — APPROVED** [review]
Thanks

File: src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/ClassFile.java [hunk 3]
```diff
@@ -123,7 +123,8 @@ public enum Version {
         V62(62, 0),   // JDK 18
         V63(63, 0),   // JDK 19
         V64(64, 0),   // JDK 20
-        V65(65, 0);   // JDK 21
+        V65(65, 0),   // JDK 21
+        V66(66, 0);   // JDK 22
```
Comment (lahodaj) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/ClassFile.java:127]: A very small nit/suggestion - it should be possible to do:
```
V66(66, 0),   //JDK 22
;
```

(i.e. ending the enum constant with `,`, and putting the semicolon on a new line.) This way adding a new constant would mean just one line addition, no modification. (The same could be done for other enums.)

File: test/langtools/tools/javac/versions/Versions.java [hunk 4]
```diff
@@ -89,7 +89,8 @@ static enum SourceTarget {
         EIGHTEEN(false,  "62.0", "18", Versions::checksrc18),
         NINETEEN(false,  "63.0", "19", Versions::checksrc19),
         TWENTY(false,    "64.0", "20", Versions::checksrc20),
-        TWENTY_ONE(false,"65.0", "21", Versions::checksrc21);
+        TWENTY_ONE(false,"65.0", "21", Versions::checksrc21),
+        TWENTY_TWO(false,"66.0", "22", Versions::checksrc21);
```
Comment (lahodaj) [inline test/langtools/tools/javac/versions/Versions.java:93]: Should there be `checksrc22` instead of `checksrc21`? Or is that done later?

**Review by lahodaj — APPROVED** [review]
javac and symbol related changes seem reasonable to me. Two minor comments added for consideration.

File: test/langtools/tools/javac/versions/Versions.java (hunk 4 above)
Comment (jddarcy) [reply test/langtools/tools/javac/versions/Versions.java]: Good catch. I have a refactoring of the test planned which will eliminate the explicit "checksrc$N" methods.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/ClassFile.java (hunk 3 above)
Comment (jddarcy) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/ClassFile.java:127]: Updated.

**Review by AlanBateman — COMMENTED** [review]
Ready to go :-)

**Review by AlanBateman — APPROVED** [review]

**Review by irisclark — APPROVED** [review]

**Pull request closed without merging**